src/Symfony/Component/HttpClient/ResponseSetMonitor.php

```diff
@@ -0,0 +1,155 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\HttpClient;
+
+use Symfony\Contracts\HttpClient\Exception\ExceptionInterface;
+use Symfony\Contracts\HttpClient\HttpClientInterface;
+use Symfony\Contracts\HttpClient\ResponseInterface;
+
+/**
+ * Monitors a set of responses and triggers callbacks as they complete.
+ *
+ * @author Nicolas Grekas <p@tchwork.com>
+ */
+final class ResponseSetMonitor implements \Countable
+{
+    private $client;
+    private $onHeaders;
+    private $onBody;
+    private $onError;
+    private $responses;
+
+    public function __construct(HttpClientInterface $client, callable $onHeaders = null, callable $onBody = null, callable $onError = null)
+    {
+        $this->client = $client;
+        $this->onHeaders = $onHeaders;
+        $this->onBody = $onBody;
+        $this->onError = $onError;
+        $this->responses = new \SplObjectStorage();
+    }
+
+    /**
+     * Adds a response to the monitored set.
+     *
+     * The response must be created with the same client that was passed to the constructor.
+     */
+    public function add(ResponseInterface $response, callable $onHeaders = null, callable $onBody = null, callable $onError = null): void
+    {
+        $this->responses[$response] = [$onHeaders, $onBody, $onError];
+        $this->tick();
+    }
+
+    /**
+     * Monitors pending responses, moving them forward as network activity happens.
+     *
+     * @param float $timeout The maximum duration of the tick
+     *
+     * @return int The number of responses remaining in the set after the tick
+     */
+    public function tick(float $timeout = 0.0): int
+    {
+        return $this->wait($timeout, false);
+    }
+
+    /**
+     * Completes all pending responses.
+     *
+     * @param float|null $idleTimeout The maximum inactivy timeout before erroring idle responses
+     */
+    public function complete(float $idleTimeout = null): void
+    {
+        $this->wait($idleTimeout, true);
+    }
+
+    /**
+     * Cancels all pending responses.
+     */
+    public function cancel(): void
+    {
+        foreach ($this->responses as $response) {
+            $response->cancel();
+        }
+
+        $this->responses = new \SplObjectStorage();
+    }
+
+    /**
+     * Returns the number of pending responses.
+     */
+    public function count(): int
+    {
+        return \count($this->responses);
+    }
+
+    public function __destruct()
+    {
+        $this->wait(null, true);
+    }
+
+    private function wait(?float $timeout, bool $errorOnTimeout): int
+    {
+        $error = null;
+        $remainingTimeout = $timeout;
+
+        if (!$errorOnTimeout && $remainingTimeout) {
+            $startTime = microtime(true);
```

Comment:
Why using a system why startTime when you can decrease remainingTimeout and check when it's <0 ?